Sources/AsyncAlgorithms/Channels/ChannelStorage.swift

```diff
@@ -0,0 +1,152 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+final class ChannelStorage<Element, Failure: Error>: @unchecked Sendable {
+  private let stateMachine: ManagedCriticalState<ChannelStateMachine<Element, Failure>>
+  private let ids = ManagedCriticalState<Int>(0)
+
+#if DEBUG
+  var onSendSuspended: (() -> Void)?
+  var onNextSuspended: (() -> Void)?
+#endif
+
+  init() {
+    self.stateMachine = ManagedCriticalState(ChannelStateMachine())
+  }
+
+  func generateId() -> Int {
+    self.ids.withCriticalRegion { ids in
+      defer { ids += 1 }
+      return ids
+    }
+  }
+
+  func send(element: Element) async {
+    let producerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    await withTaskCancellationHandler {
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
+        self.stateMachine.withCriticalRegion { stateMachine in
+          let wasCancelled = wasCancelled.withCriticalRegion { $0 }
+          guard !wasCancelled else {
+            continuation.resume()
+            return
+          }
+
+          let action = stateMachine.newProducer(continuation: continuation, element: element, producerId: producerId)
+
+          switch action {
+            case .suspend:
+#if DEBUG
+              self.onSendSuspended?()
+#endif
+              break
+            case .resumeProducer:
+              continuation.resume()
+            case .resumeProducerAndConsumer(let consumerContinuation):
+              continuation.resume()
+              consumerContinuation?.resume(returning: element)
+          }
+        }
+      }
+    } onCancel: {
+      self.stateMachine.withCriticalRegion { stateMachine in
+        wasCancelled.withCriticalRegion { $0 = true }
+        let action = stateMachine.producerWasCancelled(producerId: producerId)
+
+        switch action {
+          case .none:
+            break
+          case .resumeProducer(let continuation):
+            continuation?.resume()
+        }
+      }
+    }
+  }
+
+  func fail(error: Error) where Failure == Error {
```

Comment:
Why do we need this generic constraint here?

Reply:
That is a constraint that already exists in the current implementation of `AsyncThrowingChannel` and that I also applied on the storage. I guess as the storage is an internal implementation would could remove that constraint indeed.

Reply:
Yeah I think there is no value on having this constraint here

Reply:
I've removed the constraint.